### DISCUSSION TITLE
Eloquent eager loading, bug or intended behaviour?

### DISCUSSION THREAD

**Discussion by dododedodonl** [laravel/framework#33266] [General] [answered]
- Laravel Version: 7.16.1
- PHP Version: 7.4.3
- Database Driver & Version: mariadb 10.4.10

### Description:
When using eager loading on a model, another query's output in the relationship method changes.

### Steps To Reproduce:
Model relations:
```
User -- hasOne --> Merchant
Post -- hasMany --> Comment
Merchant -- belongsTo --> User
Comments-- belongsTo --> Post
```

Have the default user model with relationship merchant
```php
<?php

namespace App;

use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;

class User extends Authenticatable
{
    use Notifiable;

    public function merchant()
    {
        return $this->hasOne(\App\Merchant::class);
    }
}
```

Have a post model with an other relationship, depending on other data
```php
<?php

namespace App;

use Illuminate\Database\Eloquent\Model;

class Post extends Model
{
    public function comments()
    {
        return $this->hasMany(\App\Comment::class)
            ->where('priviliged', auth()->user()->merchant->priviliged);
    }
}
```

When the comments are retrieved as `Post::first()->comments`, the  code works as expected.
When the comments are retrieved as `Post::with('comments')->first()->comments`, the code does not work as expected.

I tried narrowing down the problem, and it looks like the nested query is the problem.
As a simplification I used this code.

```php
<?php

namespace App;

use Illuminate\Database\Eloquent\Model;

class Post extends Model
{
    public function comments()
    {
        $user = User::find(3); // hard code user so this is no variable
        dump($user->merchant->id); // retrieve the id to verify which data is retrieved

        return $this->hasMany(\App\Comment::class)
            ->where('priviliged', auth()->user()->merchant->priviliged);
    }
}
```

When running `Post::first()->comments`, it dumps the correct merchant id (in this case 3).
The query executed for the User->merchant relationship is `select * from merchants where merchants.user_id = ? and merchants.user_id is not null`.

When running `Post::with('comments')->first()->comments`, it dumps a different merchant id (in this case 1). Upon further inspection, all merchants are retrieved, and the first merchant is selected.
The query exectued for the User->merchant relationship is `select * from merchants`.

I do not understand enough of the eager loading mechanism to determine if this is intended behaviour or a bug.

**Comment by staudenmeir** [comment]
What's the result of `dump(auth()->user()->getAttributes(), auth()->user()->merchant->getAttributes());` in the last relationship?

**Reply by dododedodonl** [reply]
For `Post::first()->comments` the results of the 2 dumps (of seeded faker data) are:

```
array:9 [
  "id" => 3
  "name" => "Dr. Cortez Parker Jr."
  "email" => "emccullough@example.net"
  "email_verified_at" => "2020-06-18 13:31:27"
  "password" => "$2y$10$92IXUNpkjO0rOQ5byMi.Ye4oKoEa3Ro9llC/.og/at2.uheWG/igi"
  "remember_token" => "3EFVcjXHtT"
  "created_at" => "2020-06-18 13:31:27"
  "updated_at" => "2020-06-18 13:31:27"
  "deleted_at" => null
]
array:6 [
  "id" => 3
  "user_id" => 3
  "name" => "Mr. Eduardo Goldner"
  "created_at" => "2020-06-18 13:31:27"
  "updated_at" => "2020-06-18 13:31:27"
  "deleted_at" => null
]
```

For `Post::with('comments')->first()->comments` the results of the 2 dumps are:

```
array:9 [
  "id" => 3
  "name" => "Dr. Cortez Parker Jr."
  "email" => "emccullough@example.net"
  "email_verified_at" => "2020-06-18 13:31:27"
  "password" => "$2y$10$92IXUNpkjO0rOQ5byMi.Ye4oKoEa3Ro9llC/.og/at2.uheWG/igi"
  "remember_token" => "3EFVcjXHtT"
  "created_at" => "2020-06-18 13:31:27"
  "updated_at" => "2020-06-18 13:31:27"
  "deleted_at" => null
]
array:6 [
  "id" => 1
  "user_id" => 1
  "name" => "Miss Cassidy Christiansen Sr."
  "created_at" => "2020-06-18 13:31:27"
  "updated_at" => "2020-06-18 13:31:27"
  "deleted_at" => null
]
```

When I change the relationship from hasOne to hasMany, in the second case all merchants (existing in the database) are returned, while in the first case only the correct merchant is returned. Also, for the second case the where clause is not set on the querybuilder object, while it is set for the first case.

**Comment by staudenmeir** [comment]
Can you log all the executed queries for both lazy and eager loading?

**Reply by dododedodonl** [reply]
I used this comments function on the post model.
```php
public function comments()
{
    $user = User::find(3); // hard code user so this is no variable
    dump($user->merchant->id); // retrieve the id to verify which data is retrieved

    dump(auth()->user()->getAttributes());
    dump(auth()->user()->merchant->getAttributes());

    return $this->hasMany(\App\Comment::class);
}
```

Running `Post::first()->comments`, the next queries are executed:
```sql
select * from `posts` limit 1   
select * from `users` where `users`.`id` = ? limit 1   
select * from `merchants` where `merchants`.`user_id` = ? and `merchants`.`user_id` is not null limit 1   
select * from `users` where `id` = ? limit 1   
select * from `merchants` where `merchants`.`user_id` = ? and `merchants`.`user_id` is not null limit 1   
select * from `comments` where `comments`.`post_id` = ? and `comments`.`post_id` is not null 
```

Running `Post::with('comments')->first()->comments`, the next queries are executed:
```sql
select * from `posts` limit 1   
select * from `users` where `users`.`id` = ? limit 1   
select * from `merchants` limit 1   
select * from `users` where `id` = ? limit 1   
select * from `merchants` limit 1   
select * from `comments` where `comments`.`post_id` in (1)  
```

**Comment by staudenmeir** [comment]
Please share the complete `User`, `Post`, `Comment` and `Merchant` models.

**Reply by dododedodonl** [reply]
Before posting anything here, I already dummed them down as far as I could and still had the same results. I used these models for testing.

User.php
```php
<?php

namespace App;

use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;

class User extends Authenticatable
{
    use Notifiable;

    public function merchant()
    {
        return $this->hasOne(\App\Merchant::class);
    }
}
```

Post.php
```php
<?php

namespace App;

use Illuminate\Database\Eloquent\Model;

class Post extends Model
{
    public function comments()
    {
        $user = User::find(3);
        dump($user->merchant->id);

        dump(auth()->user()->getAttributes());
        dump(auth()->user()->merchant->getAttributes());

        return $this->hasMany(\App\Comment::class);
    }
}
```

Comment.php
```php
<?php

namespace App;

use Illuminate\Database\Eloquent\Model;

class Comment extends Model
{
    public function post()
    {
        return $this->belongsTo(\App\Post::class);
    }
}
```

Merchant.php
```php
<?php

namespace App;

use Illuminate\Database\Eloquent\Model;

class Merchant extends Model
{
    public function user()
    {
        return $this->belongsTo(\App\User::class);
    }
}
```

**Comment by staudenmeir** [comment]
Can you reproduce the issue on a fresh Laravel installation?

**Reply by dododedodonl** [reply]
Yes, that is what I did before posting here. The results are the same on a fresh laravel 7.16.1 install.

**Reply by staudenmeir** [reply]
Can you upload it to a GitHub repository?

**Reply by dododedodonl** [reply]
Just did that: https://github.com/dododedodonl/temp-elqouent-test

**Comment by staudenmeir** [comment, marked as answer]
The behavior is surprising, but I wouldn't consider it a bug. Relationships aren't meant to be built dynamically with additional queries.

Eager loading removes relationship constraints with `Relation::noConstraints()` and this also affects other queries that happen during this eager loading process.

Possible workarounds for your situation:

- Always load the merchant with the user:

```php
class User extends Authenticatable
{
    protected $with = ['merchant'];
}
```

- Use eager loading for the merchant:

```php
class Post extends Model
{
    public function comments()
    {
        return $this->hasMany(\App\Comment::class)
            ->where('priviliged', auth()->user()->load('merchant')->merchant->priviliged);
    }
}
```

**Reply by dododedodonl** [reply]
Okay, thank you for clarifying the process. Should this surprising behavior maybe be mentioned in the docs?

**Comment by marius-ciclistu** [comment]
This ancient issue has a solution that Laravel does not care about. More details here https://github.com/laravel/framework/issues/51825#issuecomment-2419387066

Good news is that the solution is available for laravel and lumen >=8 via external package to prevent data leakage.
